### ISSUE TITLE
pseudo-files like in bash have no equivalent

### ISSUE THREAD

**Issue by con-f-use** [xonsh/xonsh#3127]
In bash (and other shells), you can use the output of any command in a place where a file would normally be specified. The syntax would something like this:

```
┌─15-12:15[confus@jcgbx:~] $ 
└▪ echo $BASH_VERSION
4.4.19(1)-release

┌─15-12:13[confus@jcgbx:~] $ 
└▪ grep foo <(echo -e "bla laber\nschwurbel foo bar parlens\nook")
schwurbel foo bar parlens
```

Is their an equivalent in `xonsh`? If not, there should be - a least with the `bashism` xontrib enabled.

**Comment by scopatz** [comment]
Thanks for opening this @con-f-use. This is kind of three issues rolled into one, so I'll just address the top one. 

This is definitely a missing feature.  I think that most xonsh-like method of handling this would be to add a `temp`, or `t`, redirect target that would write out to a temporary file.  For example, you'd write something like:

```
grep foo $(echo -e "bla laber\nschwurbel foo bar parlens\nook" o>t)
```

When using this flag, the `$(cmd)` operator would return the file handle (int) or the string abspath to a temporary file. 

This would be great to have

**Comment by lieryan** [comment]
A couple things to note here, process substitution does not actually create a real file, but rather it works by attaching a filename to an anonymous pipe using either the /def/fd/ filesystem or mkfifo. A real file cannot actually be used to correctly implement process substitution. Process substitution being a pipe have a number of consequences:

- The substituted process only keeps a small buffer and will be blocked if the main command isn't reading from the file. This prevents the substituted process from causing the system to run out of disk space when the output is large, and it'll allow process substitution to work correctly if the substitute process is a command that produces infinite output

- The substitute file isn't seekable, since it's actually a pipe, not a real file.

- The output of the substituted file never touches persistent storage, which can be important when working with sensitive data

- Like a pipe and file redirection, there's a direction to the substituted process, you can either specify a readable substitute `consumer <(producer)` or a writable substitute `producer >(consumer)`.

----

As for syntax, personally, after thinking about this for the last few days, I can't really find a better syntax than the bash syntax. Pretty much the only drawback of bash syntax is that it's not Google-able. If I were to design a more search engine friendly syntax for it though, it might look something along the line of this:

    >>> foo --input !(bar).reader --output !(baz).writer 

Though frankly speaking, once you're familiar with regular input/output file redirection syntax, the arrow created by the square bracket actually makes for a very intuitive directionality of process substitution as well. Using explicit words like stdin/stdout actually may make it harder to correctly use which direction the data goes.

**Comment by AstraLuma** [comment]
You discussed having an `<@()` form the other day, where it would take a python expression that produced a file-like/buffer/iterable/etc and feed it into the STDIN of a process.

**Comment by hg42** [comment]
it might look like something about compatibility to foreign shells, but it is really about the functionality, which means constructing a network of processes that talk to each other.

E.g. you could have a command `mixer` that reads multiple inputs at once (e.g. with a thread for each) and output each incoming line.

You then can do something like this:
```
mixer <(actor_1) <(actor_2) <(actor_3) | do_something_with_lines >(reactor_1) >(reactor_2)
```

note that each line will be processed immediately.

**Comment by hg42** [comment]
well, I'm coming from zsh (with much more features than bash and also cleaner) and only looking for alternatives, that blend the conciseness of shell scripts with the structure of real languages, and also able to pipe object collections (with lines being only a variant).
Every shell script starts with a one liner (so it must be capable of doing such things effectively) and then grows into programs over time if the script is used more often.

My test for features I often use, were not really convincing...e.g.

```
for line in !(xonfig styles):
  print(line)
```

did not work like expected, while

```
for line in !(ls):
  print(line)
```
worked.

Also including `\n` at the end of each line is counter intuitive (I know there is a discussion about that).

One of the features is in this thread, using the output of a command as a filename (which means using a named pipe).

btw. shortcuts should be just like that, they should only give a more concise way of a more verbose way.
So <(command) would be the shortcut for !(command).reader
I think the shortcut syntax should regular and extensible without changing a parser. This way people can experiment easily and also cover special cases in their scripts.

**Comment by slacksystem** [comment]
I wonder is there anyone working on this issue? This is a feature I use regularly in bash and zsh and it would be great if it could be implemented here. I would offer to take a stab at it myself but I think it's probably far beyond my current skills

**Comment by anki-code** [comment]
1. I think a good start to this issue will be a clear description of `<()` syntax and use cases.

2. As I understnad (may be wrong) `cmd1 <(cmd2)` is the similar with `cmd2 | cmd1`. From my knowledge of xonsh internals I think it's possible to implement on pipes level:

    https://github.com/xonsh/xonsh/blob/d55c5c7370b0ed91b0451f7e359d329855e2a36e/xonsh/procs/specs.py#L1034-L1039

3. Another question is about changes to parser - I can't help here.

4. Finally based on count of thumbs up for this issue I think it's not so popular thing. It will be cool to have most useful examples here.

**Comment by slacksystem** [comment]
I think someone's people forget to do the thumbs up to things they think would be helpful. I just realized after your message that I hadn't reacted (have now), so I don't think that necessarily means it isn't worth doing

In my eyes, `cmd2 | cmd1` is very different from `cmd1 <(cmd2)`. You can't do the latter with regular pipes. In the first one, `cmd1` is a command that accepts _standard input_ that it gets from a pipe, but in the desired case `cmd1` takes a _file_ as an argument, but might not accept standard input, so we need to pass the output of cmd2 as a pseudofile into cmd1, without saving the output to a file. 

Examples of usefulness:

```
diff <(fd '.' -t f dir1 -x cat) <(fd '.' -t d dir1 -x cat)
```
This would compare all of the contents of the files in `dir1` recursively with the contents of the files in `dir2`.
```
ls | tee >(grep 'txt$') >(wc -l) > /dev/null
```
Sends the output of the first command into two different commands .
```
sort -m <(sort file1.txt) <(sort file2.txt) > merged.txt
```
Sort two files individually and then merge sort them together 
```
diff -y <(tree -fi dir1) <(tree -fi dir2)
```
Diff the directory trees of two directories

Basically `cmd1 <(cmd2)` makes a named pipe and pass it to the `cmd1` as if it were a file, sending the output of `cmd2` through the pipe, and `cmd1 >(cmd2)` makes a named pipe and passes it to `cmd2` as a file, sending the output of `cmd1`.

I could try to write a function that handles these processes if that would be helpful, but I wouldn't know how to integrate it as syntax in the lexer.

**Comment by anki-code** [comment]
On command decorator level `<()` can be implemented like this:

```xsh
class DecoratorAliasTmpFile(@.imp.xonsh.procs.specs.DecoratorAlias):
    @staticmethod
    def stream_lines_to_tmp_file(lines):
        """
        Instead of returning the output we can write it to the tmp file and return path to file.
        Note:
        1. May be instead of writing the lines to file manually we can just add `> tmp_file` to last `spec`
           or use OS named pipes.
        2. No deletion of tmp file. Suggestions? May be we need to have a way to go upstream
           command and have an ability to decorate the initial command ending. 
        """
        with @.imp.tempfile.NamedTemporaryFile(dir="/tmp", delete=False, mode="w") as tmp_file:
            for l in lines:
                tmp_file.write(l + __xonsh__.imp.os.sep)
            return tmp_file.name
    
    def decorate_spec(self, spec):
        """
        Just set the output_format in DecoratorAlias.
        Good to learn:
          * https://xon.sh/tutorial.html#decorator-aliases
          * https://github.com/anki-code/xontrib-dalias
        """
        spec.output_format = self.stream_lines_to_tmp_file

aliases['@asfile'] =  DecoratorAliasTmpFile()
```
So:
```xsh
mkdir -p /tmp/tst && cd /tmp/tst && touch 123 .123

echo $(@asfile echo 123)
# /tmp/tmpn0w_d_qg

diff $(@asfile ls -a) $(@asfile ls)
# 1,4c1
# < .
# < /..
# < /.123
# < /123
# ---
# > 123

diff $(@asfile ls -a) $(ls | @asfile grep 1)  # add alias decorator to the last command in pipe
```

If it's working to you it will be cool to test and add to [xontrib-dalias](https://github.com/anki-code/xontrib-dalias). PR is welcome!

**Comment by slacksystem** [comment]
I think named pipes are much better for this than temp files. At this point I have:
```xsh
class DecoratorAliasAsPath(@.imp.xonsh.procs.specs.DecoratorAlias):
    @staticmethod
    def stream_lines_to_named_pipe(lines):
        """
        Instead of returning the output we can write it to a named pipe and return the path.
        Using a named pipe avoids writing to disk and is more secure than using a temporary file.
        """
        with tempfile.TemporaryDirectory() as temp_dir:
            fifo_path = os.path.join(temp_dir, 'temp_fifo')
            with open(fifo_path, 'w') as fifo:
                for line in lines:
                    fifo.write(line + '\n')
            return fifo_path


    
    def decorate_spec(self, spec):
        """
        Just set the output_format in DecoratorAlias.
        Good to learn:
          * https://xon.sh/tutorial.html#decorator-aliases
          * https://github.com/anki-code/xontrib-dalias
        """
        spec.output_format = self.stream_lines_to_named_pipe

aliases['@asfile'] =  DecoratorAliasAsPath()
```
The problem is, the named pipe doesn't stay open long enough for the enclosing process to read it. It's a similar issue to cleaning up temporary files, in that we need to let the enclosing process finish before cleanup / closing. I don't know how to get information about the enclosing process, is that something I could get from `spec`? Is `spec` for the command that is being modified by the decorator alias only, or is it the spec for the full command in which the dalias'ed command is called?

**Comment by anki-code** [comment]
If you want to dive into this take a look [how pipes work](https://github.com/xonsh/xonsh/issues/3127#issuecomment-3538900467). Using decorator alias is just a way how to experiment with spec (but I like how it looks i.e. `cmd $(@asfile cmd)`).

To implement this feature (`<()`) completely we need to play with named pipes in Python code. I mean we have [the code that transform command to specifications](https://github.com/xonsh/xonsh/blob/d55c5c7370b0ed91b0451f7e359d329855e2a36e/xonsh/procs/specs.py#L1028). When you run `echo 1 | grep 1 > /tmp/file` there is python code that creates `p1 = spec("echo 1")`, `p2 = spec(grep 1)` and then [set](https://github.com/xonsh/xonsh/blob/d55c5c7370b0ed91b0451f7e359d329855e2a36e/xonsh/procs/specs.py#L1038-L1039) `p2.stdin = p1.stdout` and then [set](https://github.com/xonsh/xonsh/blob/d55c5c7370b0ed91b0451f7e359d329855e2a36e/xonsh/procs/specs.py#L200) `p2.stdout = file('/tmp/file')`. Are you understanding? 
So now to implement the similar logic for the requested new feature we need to implement new machinery. For a command `diff $(@asfile echo 1) $(@asfile echo 2)` (three processes) we need to create named pipes and somehow manage the processes on Python. In fact we need to implement bash `<()` feature in Python.

**Comment by anandghegde** [comment]
I'd like to implement this. @anki-code asked for a clear syntax and use cases, so here they are.

Syntax (subprocess mode only):
- `cmd <(inner)`: xonsh runs `inner` and passes `cmd` a path it can read from. `cmd >(inner)`: `cmd` gets a path to write to, and `inner` reads from it.
- As in bash, no space between `<`/`>` and `(`. On current `main`, `![cat <(echo 1)]` and `![echo 1 >(cat)]` are both SyntaxErrors, so nothing that works today changes. In Python mode, `a <(b)` stays a comparison. The lexer already tracks the mode (see the #4994 fix in `handle_redirect`), so it emits the new tokens only in subprocess mode. `base.py` gets a `subproc_arg_part` rule next to `$(...)`.

Runtime:
- POSIX only. `mkfifo` inside a private `mkdtemp()` directory, with the inner command's stdout (or stdin) connected to it. A FIFO rather than `/dev/fd/N`, because it works for callable aliases too and needs no fd inheritance through `Popen`. Windows raises a clear error.
- Lifetime: the substitutions belong to the enclosing command. When its `CommandPipeline` ends, the FIFO and directory are removed and the inner process is waited on, or terminated if nothing ever opened the path. That covers the cleanup problem @slacksystem ran into.

Use cases: `diff <(ls a) <(ls b)`, `sort -m <(sort f1) <(sort f2)`, `ls | tee >(grep txt) >(wc -l)`.

Questions:
1. Is subprocess-only, whitespace-sensitive syntax acceptable?
2. Should the opt-in RD parser (`XONSH_RD_PARSER`) be covered in the same PR?
3. Should the first PR include `>()`, or `<()` only?

Happy to open a PR once the syntax is agreed.

**Comment by slacksystem** [comment]
Just my opinion, but the syntax you described sounds good. And the features are paired so I think it makes sense to include them both in the same PR, but definitely the same release if not the same PR.

**Comment by anandghegde** [comment]
Thanks @slacksystem - agreed, I'll do both `<()` and `>()` in the same PR.

@anki-code before I start: are you OK with the subprocess-only, whitespace-sensitive syntax above, and would you rather I cover the RD parser (`XONSH_RD_PARSER`) in the same PR or leave it for a follow-up?

**Comment by anki-code** [comment]
Personally I don't like `>()`/`<()` syntax. Yes, it's POSIX but parsing in Python context could produce side effects. But it's not the key thing in this feature.
When I tried to implement this feature I faced with complex implementation. We need to clear allocated FDs *after* the command was executed and it's not trivial without deep refactoring of executor. We need some post-work after the command was finished. First of all try to dive into it and if you're using LLM try to talk about that as deep as possible.
